src/frequenz/sdk/timeseries/_formula_engine/_formula_generators/_consumer_power_formula.py

```diff
@@ -50,95 +50,141 @@ def generate(self) -> FormulaEngine[Power]:
 
-        if len(grid_successors) == 1:
-            grid_meter = next(iter(grid_successors))
-            if grid_meter.category != ComponentCategory.METER:
-                raise RuntimeError(
-                    "Only grid successor in the component graph is not a meter."
-                )
-            return self._gen_with_grid_meter(builder, grid_meter)
-        return self._gen_without_grid_meter(builder, grid_successors)
+        component_graph = connection_manager.get().component_graph
+        if all(
+            successor.category == ComponentCategory.METER
+            and not component_graph.is_battery_chain(successor)
+            and not component_graph.is_chp_chain(successor)
+            and not component_graph.is_pv_chain(successor)
+            and not component_graph.is_ev_charger_chain(successor)
+            for successor in grid_successors
+        ):
+            return self._gen_with_grid_meter(builder, grid_successors)
+
+        return self._gen_without_grid_meter(builder, self._get_grid_component())
 
     def _gen_with_grid_meter(
         self,
         builder: ResampledFormulaBuilder[Power],
-        grid_meter: Component,
+        grid_meters: Set[Component],
     ) -> FormulaEngine[Power]:
         """Generate formula for calculating consumer power with grid meter.
 
         Args:
             builder: The formula engine builder.
-            grid_meter: The grid meter component.
+            grid_meters: The grid meter component.
 
         Returns:
             A formula engine that will calculate the consumer power.
         """
         component_graph = connection_manager.get().component_graph
-        successors = component_graph.successors(grid_meter.component_id)
 
-        builder.push_component_metric(grid_meter.component_id, nones_are_zeros=False)
+        def non_consumer_component(component: Component) -> bool:
+            """
+            Check if a component is not a consumer component.
+
+            Args:
+                component: The component to check.
 
-        for successor in successors:
+            Returns:
+                True if the component is not a consumer component, False otherwise.
+            """
             # If the component graph supports additional types of grid successors in the
             # future, additional checks need to be added here.
-            if (
-                component_graph.is_battery_chain(successor)
-                or component_graph.is_chp_chain(successor)
-                or component_graph.is_pv_chain(successor)
-                or component_graph.is_ev_charger_chain(successor)
-            ):
+            return (
+                component_graph.is_battery_chain(component)
+                or component_graph.is_chp_chain(component)
+                or component_graph.is_pv_chain(component)
+                or component_graph.is_ev_charger_chain(component)
+            )
```

Comment:
Would it be possible to consolidate `non_consumer_component` and `consumer_component`  into only one internal helper method in order to avoid code duplication?

Reply:
Actually I though about it and decided against it. The reason is that we'd have a function with some parameters that change the behavior, which I find harder to reason about, then what we have here.